### PR TITLE
lesson 14

### PR DESCRIPTION
**Pull request by DenysKopchenov**
-implemented LinkedList for product versioning with required methods
-created tests for VersioningLinkedList
-add equals and hashCode for Products
-created Product comparator first comparing prices from high to low, then titles in alphabetical order, then count from low to high

### CODE REVIEW

File: src/main/java/com/alevel/lesson10/shop/service/VersioningLinkedList.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        } else {
+            f.prev = newNode;
+        }
+        size++;
+        versions.add(newNode.version);
+    }
+
+    public void addFirst(E e) {
+        linkFirst(e);
+    }
```
Comment (Fabolo12) [inline src/main/java/com/alevel/lesson10/shop/service/VersioningLinkedList.java:36]: А зачем тут разделять на разные методы?

File: src/main/java/com/alevel/lesson10/shop/service/VersioningLinkedList.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        return size;
+    }
+
+    @Override
+    public Iterator<E> iterator() {
+        return new VersioningLinkedListIterator();
+    }
+
+    @Override
+    public void forEach(Consumer action) {
```
Comment (Fabolo12) [inline src/main/java/com/alevel/lesson10/shop/service/VersioningLinkedList.java:130]: `public void forEach(Consumer<? super E> action) {`

**Review by Fabolo12 — APPROVED** [review]

File: src/main/java/com/alevel/lesson10/shop/service/VersioningLinkedList.java (hunk 2 above)
Comment (DenysKopchenov) [reply src/main/java/com/alevel/lesson10/shop/service/VersioningLinkedList.java:130]: Исправлю

File: src/main/java/com/alevel/lesson10/shop/service/VersioningLinkedList.java (hunk 1 above)
Comment (DenysKopchenov) [reply src/main/java/com/alevel/lesson10/shop/service/VersioningLinkedList.java:36]: Наверное не было смысла, исправлю

**Pull request merged**